libgadget/fof.c

```diff
@@ -386,34 +397,38 @@ static void
 fofp_merge(int target, int other, TreeWalk * tw)
 {
     /* this will lock h1 */
-    int h1 = HEADl(-1, target, tw);
-    /* stop looking if we find h1 along the path (because it is already owned by us) */
-    int h2 = HEADl(h1, other, tw);
-
-    if(h2 == -1) {
-        /* h1 is along the path of h2, already merged.  **/
-        /* h1 must be the root of other and target both */
-        //printf("unlocking %d by %d in merge\n", h1, omp_get_thread_num());
-        update_root(target, h1, tw);
-        update_root(other, h1, tw);
-        unlock_particle(h1);
-        return;
-    }
-
-    /* h2 as a sub-tree of h1 */
-    FOF_PRIMARY_GET_PRIV(tw)->Head[h2] = h1;
-
-    /* update MinID of h1 */
-    if(HaloLabel[h1].MinID > HaloLabel[h2].MinID)
+    int * Head = FOF_PRIMARY_GET_PRIV(tw)->Head;
+    int h1, h2;
+
+    do {
+        h1 = HEADl(-1, target, -1, Head);
+        /* stop looking if we find h1 along the path (because it is already owned by us) */
+        h2 = HEADl(h1, other, h1, Head);
+        /* We had a lock already taken on h2 by another thread.
+         * We need to unlock h1 and retry to avoid deadlock loops.*/
+        if(h2 == -2)
+            unlock_particle(h1);
+    } while(h2 == -2);
+
+    if(h2 >=0)
     {
-        HaloLabel[h1].MinID = HaloLabel[h2].MinID;
-        HaloLabel[h1].MinIDTask = HaloLabel[h2].MinIDTask;
+        /* h2 as a sub-tree of h1 */
+        Head[h2] = h1;
+
+        /* update MinID of h1 */
+        if(HaloLabel[h1].MinID > HaloLabel[h2].MinID)
+        {
+            HaloLabel[h1].MinID = HaloLabel[h2].MinID;
+            HaloLabel[h1].MinIDTask = HaloLabel[h2].MinIDTask;
+        }
+        //printf("unlocking %d by %d in merge\n", h2, omp_get_thread_num());
+        unlock_particle(h2);
     }
-    //printf("unlocking %d by %d in merge\n", h2, omp_get_thread_num());
-    unlock_particle(h2);
 
-    update_root(target, h1, tw);
-    update_root(other, h1, tw);
+    /* h1 must be the root of other and target both:
```

Comment:
If h1 is locked, then any object on path from target to h1 can be splayed.

After h2 is unlocked, the root of any object under h2 becomes h1. As h1 is still locked, we can also safely splay any path along other to h1. That's why the second loop is after unlocking of h2. 

Does this make sense? If so, add a comment so we don't miss it next time.

Reply:
Yes, I see - there is no need to hold the snapshot of h2 because everything looking at it will reach h1 anyway. I will add a comment.